librz/core/devirtualize_cxx.c

```diff
@@ -0,0 +1,956 @@
+// SPDX-FileCopyrightText: 2025 tushar3q34 <tushar3q34@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_util.h>
+#include <rz_analysis.h>
+#include <rz_core.h>
+#include "core_private.h"
+
+/**
+ * \file
+ * Partially uses ESIL uplifting but will be ported to RzIL
+ */
+
+/**
+ * \brief Used to describe current state of tainting
+ */
+typedef enum rz_taint_devirt_t {
+	RZ_TAINT_MODE_OFF, ///< No tainting in this mode
+	RZ_TAINT_MODE_RED, ///< Mode when `new` is identified but constructor is yet to be identified
+	RZ_TAINT_MODE_BLUE, ///< Mode after constructor is also identified, write to variable is yet to be identified
+} RzTaintDevirt;
+
+typedef struct var_t {
+	char *name; ///< for now, checks varibles discovered by ESIL
+	ut64 stack_addr; ///< address of the stack where variable is stored according to RzIL VM
+	ut64 instr_addr; ///< address of instruction which writes to variable
+	RzList /*<char *>*/ *class_names; ///< Single variable might store multiple classes based on conditionals
+	RzList /*<ut64 *>*/ *object_addr; ///< Address of allocated object which stores vtable ptr
+} CppVariable;
+
+typedef struct rz_taint_state_t {
+	RzTaintDevirt taint; ///< current taint mode
+	RzAnalysisOp *op; ///< analysis instruction op
+	ut64 addr; ///< current instruction address
+	char *class_name; ///< class name to be stored in variable
+	RzCppVariableBook *var_book; ///< collection of all stack variables
+	HtSP /*<char *,char *>*/ *method_class_map; ///< maps methods to class name
+} RzTaintState;
+
+typedef struct rz_virtual_calls_t {
+	HtUP /*<ut64,RzSetS *>*/ *virt_calls; ///< hash map of instruction address and class method name
+} RzVirtualCalls;
+
+#define RZ_TAINT_VALUE 0x1A2B3C
+
+/**
+ * \return list of `RzAnalysisFunction *` of all allocator i.e. `new` functions
+ */
+RzList /*<RzAnalysisFunction *>*/ *list_new_fcns(RzCore *core) {
+	RzList *list = rz_analysis_function_list(core->analysis);
+	if (!list) {
+		return NULL;
+	}
+	RzListIter *it;
+	RzAnalysisFunction *fcn;
+	RzList *ret_list = rz_list_new();
+	rz_list_foreach (list, it, fcn) {
+		char *res = strstr(fcn->name, "sym.imp.operator_new");
+		if (res != NULL) {
+			rz_list_push(ret_list, it->elem);
+		}
+	}
+	return ret_list;
+}
+
+static void rz_taint_vm_step(RzCore *core) {
+	if (!rz_core_il_step(core, 1)) {
+		return;
+	}
+
+	if (!core->analysis || !core->analysis->il_vm) {
+		return;
+	}
+
+	RzILVM *vm = core->analysis->il_vm->vm;
+
+	void **it;
+	RzILEvent *evt;
+
+	rz_config_set_cb(core->config, "rzil.step.events.read", "true", NULL);
+	rz_config_set_cb(core->config, "rzil.step.events.write", "true", NULL);
+
+	rz_pvector_foreach (vm->events, it) {
+		evt = *it;
+		switch (evt->type) {
+		case RZ_IL_EVENT_MEM_WRITE: {
+			ut64 new_val = rz_bv_to_ut64(evt->data.mem_write.new_value);
+			if (new_val == RZ_TAINT_VALUE) {
+				return;
+			}
+		}
+		default: {
+			break;
+		}
+		}
```

Comment:
maybe just make an `if (evt->type != RZ_IL_EVENT_MEM_WRITE) continue;` otherwise
```suggestion
		switch (evt->type) {
		case RZ_IL_EVENT_MEM_WRITE: {
			ut64 new_val = rz_bv_to_ut64(evt->data.mem_write.new_value);
			if (new_val == RZ_TAINT_VALUE) {
				return;
			}
			break;
		}
		default:
			break;
		}
```

Reply:
Second the `if` check over `switch`